spacy/pipeline/attributeruler.py

```diff
@@ -0,0 +1,252 @@
+import srsly
+from typing import List, Dict, Union, Iterable, Any
+from pathlib import Path
+
+from .pipe import Pipe
+from ..errors import Errors
+from ..language import Language
+from ..matcher import Matcher
+from ..symbols import IDS
+from ..tokens import Doc, Span
+from ..tokens._retokenize import normalize_token_attrs, set_token_attrs
+from ..vocab import Vocab
+from .. import util
+
+
+MatcherPatternType = List[Dict[Union[int, str], Any]]
+
+
+@Language.factory(
+    "attribute_ruler",
+    assigns=[],
+    default_config={},
+    scores=[],
+    default_score_weights={},
+)
+def make_attribute_ruler(
+    nlp: Language, name: str, pattern_dicts: Iterable[Dict] = tuple()
+):
+    return AttributeRuler(nlp.vocab, name, pattern_dicts=pattern_dicts)
```

Comment:
```suggestion
    nlp: Language, name: str
):
    return AttributeRuler(nlp.vocab, name)
```

We probably shouldn't allow the `pattern_dicts` to be passed in via the config – otherwise, you'll end up with those values in your `config.cfg` (if you set them on initialization with `nlp.add_pipe`).

Instead, I think the recommended usage should be:

```python
ruler = nlp.create_pipe("attribute_ruler")
ruler.add(patterns)
```

Reply:
This format may be a bad idea, but I think you need to be able to provide an asset containing patterns in the config somehow, similar to how you used to provide a tag map?

Reply:
Yeah, fair point. 

I think this ties into a more general question of how we can (and whether we should) find a way to enforce the following:

```python
# Bad
nlp.create_pipe("attribute_ruler", config={"patterns": [[{"TEXT": "foo"}]]})

# Good
nlp.create_pipe("attribute_ruler", config={"patterns": {"@assets": "my_patterns"}})
```

Reply:
Is there a technical reason to want to disallow patterns to be specified in the config? Is there a disadvantage other than potentially having a config that's long and hard to read?

And looking at the APIs, whatever we choose here should also be used for `EntityRuler`.

Reply:
In the case of the patterns, it's mostly just that it clogs up the config and once it's in there, you can't really get it out. And we don't want people to get the idea that this is how their patterns should be stored – especially since they can get quite verbose. But I'd be willing to accept this as an unfortunate side-effect.

Where it gets trickier is when objects are JSON-serializable, because then you'd end up with their serialized version in your config, not the actual reference to the object. For example:

```python
from spacy.lookups import Lookups
from thinc.api import Config

config = Config({"config": {"lookups": Lookups()}})  
print(config.to_str())
```